### PR TITLE
ICU-13883 add remove org modal

### PR DESCRIPTION
**Pull request by lisbet-alvarez**
✅ Closes: https://hashicorp.atlassian.net/browse/ICU-13883

# Description

Add a modal for when a user deselects a singular org with projects. 
Add a separate modal for when a user deselects all and orgs have projects attached to them.

## Screenshots (if appropriate)

https://github.com/user-attachments/assets/8ca51091-58c6-4320-af79-fc88239622d6

## How to Test

1. Make sure you have an org created that has various projects
2. Go to roles page -> click on a role
3. Click "manage" dropdown -> "manage scopes" -> "manage custom scopes"
4. For the Org with many projects click the link under "Projects selected"
5. Select and save a couple projects.
6. After saving, you should now be on the "manage custom scopes" page again.
7. If you deselect the org with projects then the modal should appear.
8. click "Remove org and projects" and now project count should be zero
9. Click "save" and scopes tab table should not have those projects

## Checklist

- [ ] I have added before and after screenshots for UI changes
- [ ] I have added JSON response output for API changes
- [ ] I have added steps to reproduce and test for bug fixes in the description
- [X] I have commented on my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: addons/core/translations/resources/en-us.yaml [hunk 1]
```diff
@@ -627,6 +629,9 @@ role:
       descendants:
         label: Add all descendants
         help: Add all of the orgs and the projects in the orgs that are underneath this scope.
+    remove-org:
+      title: Remove projects under this org?
+      description: Do you want to remove the projects under this org from this role? Or do you just want to remove the org from this role?
```
Comment (cameronperera) [inline addons/core/translations/resources/en-us.yaml:634]: I could be reading it wrong but when I read the first question, it feels to me like we can remove just the projects but keep the org selected, not remove the projects AND the org.

File: ui/admin/app/routes/scopes/scope/roles/role/manage-scopes/manage-custom-scopes.js [hunk 2]
```diff
@@ -7,7 +7,7 @@ import Route from '@ember/routing/route';
 import { action } from '@ember/object';
 import { inject as service } from '@ember/service';
 import { TYPE_SCOPE_PROJECT } from 'api/models/scope';
-
+import { TrackedObject } from 'tracked-built-ins';
```
Comment (cameronperera) [inline ui/admin/app/routes/scopes/scope/roles/role/manage-scopes/manage-custom-scopes.js:10]: small nitpick: Could we add a space between the imports and class declaration?

File: ui/admin/tests/acceptance/roles/global-scope-test.js [hunk 3]
```diff
@@ -509,6 +514,72 @@ module('Acceptance | roles | global-scope', function (hooks) {
...
+
+    await click(SAVE_BTN_SELECTOR);
+
+    assert.strictEqual(currentURL(), urls.manageScopes);
+    assert.dom(BUTTON_ICON_SELECTOR).isVisible();
+
+    await click(SAVE_BTN_SELECTOR);
+
+    assert.strictEqual(currentURL(), urls.roleScopes);
+    assert.strictEqual(findAll(TABLE_ROW_SELECTOR).length, 0);
```
Comment (cameronperera) [inline ui/admin/tests/acceptance/roles/global-scope-test.js:580]: I think we can do the same thing here with `assert.dom(TABLE_ROW_SELECTOR).exist({ count: 0 })`. I know we have used it in past work, however I can see that it isn't as clear. It makes me wonder if `doesNotExist()` would also work and be more clear.
```suggestion
    assert.dom(TABLE_ROW_SELECTOR).exist({ count: 0 });
```

File: ui/admin/app/routes/scopes/scope/roles/role/manage-scopes/manage-custom-scopes.js (hunk 2 above)
Comment (lisbet-alvarez) [reply ui/admin/app/routes/scopes/scope/roles/role/manage-scopes/manage-custom-scopes.js:10]: of course!

File: ui/admin/app/components/form/role/manage-custom-scopes/index.js [hunk 4]
```diff
@@ -5,27 +5,57 @@
 
 import Component from '@glimmer/component';
 import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
 
 export default class FormRoleManageCustomScopesIndexComponent extends Component {
+  // =attributes
+
+  @tracked showRemoveOrgModal = false;
+  @tracked selectedOrg;
```
Comment (ZedLi) [inline ui/admin/app/components/form/role/manage-custom-scopes/index.js:14]: Is the boolean needed if you are also keeping track of the `selectedOrg`?

File: ui/admin/app/components/form/role/manage-custom-scopes/index.js [hunk 5]
```diff
@@ -5,27 +5,57 @@
...
+        // Add the org id if it was selected and it doesn't exist as a grant scope.
         role.grant_scope_ids = [...role.grant_scope_ids, key];
       } else if (!isSelected && includesId) {
+        // Remove org id if it was deselected and it does exist as a grant scope.
         role.grant_scope_ids = role.grant_scope_ids.filter(
           (item) => item !== key,
         );
+        // If the org id was deselected and has selected projects then trigger the modal.
+        const selected = projectTotals[key]?.selected;
+        if (selectionKey === key && selected?.length) {
```
Comment (ZedLi) [inline ui/admin/app/components/form/role/manage-custom-scopes/index.js:38]: I have two questions:
1. I didn't look too closely before but seeing this key logic and looking at the docs I'm now curious, why did we have to iterate over all the `selectableRowsStates` when we get the `selectionKey` for each row? Why was it necessary to iterate over all the rows each time a row is being checked or unchecked? Is it to handle the `all` rows check option?
2. If a user clicks the all checkbox, do we need to handle removing all projects if a user unselects all orgs with it?

Comment (lisbet-alvarez) [reply ui/admin/app/components/form/role/manage-custom-scopes/index.js:38]: 1. Now that you point it out, there is a more efficient way of doing this by separating the cases that use `all`. I'll experiment with it more.
2. I think that's another good point and i'll check in with design.

**Review by cameronperera — APPROVED** [review]

File: ui/admin/app/components/form/role/manage-custom-scopes/index.js (hunk 4 above)
Comment (lisbet-alvarez) [reply ui/admin/app/components/form/role/manage-custom-scopes/index.js]: good point, the bool has been removed.

File: ui/admin/app/components/form/role/manage-custom-scopes/index.js (hunk 5 above)
Comment (lisbet-alvarez) [reply ui/admin/app/components/form/role/manage-custom-scopes/index.js]: Update: a new modal has been added for the case in which the user clicks the all checkbox. Thank u for the feedback.

File: ui/admin/app/components/form/role/manage-custom-scopes/index.js [hunk 6]
```diff
@@ -10,52 +10,117 @@ import { tracked } from '@glimmer/tracking';
...
-    selectableRowsStates.forEach((row) => {
-      const { isSelected, selectionKey: key } = row;
-      const includesId = role.grant_scope_ids.includes(key);
-      if (isSelected && !includesId) {
-        // Add the org id if it was selected and it doesn't exist as a grant scope.
-        role.grant_scope_ids = [...role.grant_scope_ids, key];
-      } else if (!isSelected && includesId) {
-        // Remove org id if it was deselected and it does exist as a grant scope.
+
+    const addOrRemoveValues = (add, remove, orgId) => {
```
Comment (ZedLi) [inline ui/admin/app/components/form/role/manage-custom-scopes/index.js:38]: Was it intentional that selecting orgs automatically selects all projects underneath it now? Before this change it didn't auto select any projects and you had to actually go in now.

It seems like now there's no easy way to just select only orgs?

Comment (lisbet-alvarez) [reply ui/admin/app/components/form/role/manage-custom-scopes/index.js:38]: yeah it was intentional, originally i had asked about that change being a modal in this [slack thread](https://hashicorp.slack.com/archives/C06LR1KKN4E/p1723568423341309).

File: ui/admin/app/components/form/role/manage-custom-scopes/index.js (hunk 5 above)
Comment (lisbet-alvarez) [reply ui/admin/app/components/form/role/manage-custom-scopes/index.js]: @ZedLi tagging for visibility to the comment above

File: ui/admin/app/components/form/role/manage-custom-scopes/index.js (hunk 6 above)
Comment (ZedLi) [reply ui/admin/app/components/form/role/manage-custom-scopes/index.js:38]: So there's no way to just select all the orgs only?

Comment (lisbet-alvarez) [reply ui/admin/app/components/form/role/manage-custom-scopes/index.js:38]: Nope, not on the manage custom scopes page. The only way would be to select the org save it (it auto selects projects) then go into the manage projects page and deselect all projects. Would you prefer if i move this specific change to a different PR?

Comment (lisbet-alvarez) [reply ui/admin/app/components/form/role/manage-custom-scopes/index.js:38]: changes around this have been removed! thank u for the feedback!

File: ui/admin/app/components/form/role/manage-custom-scopes/index.js [hunk 7]
```diff
@@ -5,27 +5,108 @@
...
+      this.toggleRemoveOrgModal();
+    }
+  }
+
+  /**
+   * Toggles the modal to remove an org and projects off.
+   */
+  @action
+  toggleRemoveOrgModal() {
+    this.selectedOrg = [];
```
Comment (ZedLi) [inline ui/admin/app/components/form/role/manage-custom-scopes/index.js:102]: Should this be an empty string instead of array?

**Review by ZedLi — APPROVED** [review]
Works great!

File: ui/admin/app/components/form/role/manage-custom-scopes/index.js (hunk 7 above)
Comment (lisbet-alvarez) [reply ui/admin/app/components/form/role/manage-custom-scopes/index.js:102]: oh yup good catch, i was experimenting with something and forgot to change it back.

**Pull request merged**